### PR TITLE
fix(button): restore disabled state set from outside when loading sta…

### PR DESCRIPTION
**Pull request by EndzeitBegins**
…te changes to default

Signed-off-by: EndzeitBegins <16666115+EndzeitBegins@users.noreply.github.com>

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] clarity.design website / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #129

## What is the new behavior?

When changing to the `loadingState` to default the `disabled` value set from outside is restored, whether is was set before or while the `loadingState` was non-default.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Other information

My solution includes replacing the field `disabled` in `CdsBaseButton` with a corresponding getter and setter in order to override its behavior in `CdsButton`. This in turn adds a private field `_disabled` to `CdsBaseButton`. To avoid collisions I renamed `_disabled` to `_disabled_from_outside` in `CdsButton`. 
According to `npm run public-api:check` this results in changes to the "public" API, even though only private fields were added / renamed and the field `disabled` replaced with getter / setter.

### CODE REVIEW

File: projects/core/src/button/button.element.spec.ts [hunk 1]
```diff
@@ -279,17 +281,62 @@ describe('button element', () => {
...
+      expect(component.disabled).toBeTruthy();
+
+      component.loadingState = ClrLoadingState.default;
+      await componentIsStable(component);
+      expect(component.disabled).toBeTruthy();
+    });
+
+    it('should go to enabled set while loadingState was non-default, when it changes back', async () => {
+      await componentIsStable(component);
+      component.disabled = true; // should NOT be restored, as it is changed while in non-default loadingState
```
Comment (ashleyryan) [inline projects/core/src/button/button.element.spec.ts:306]: I'm confused by this test, by default the loadingState is the default loading state

**Comment by ashleyryan** [comment]
Thanks for taking a stab at this, I'm reviewing and looking through it now. I don't love the idea of completely disregarding the value attempting to be set in the setter.

We already had the `_disabled` hidden variable that was tracking the manually set disabled status. And  uses it to set the disabled value when the loading state changes back to default. I'm trying to first understand why this isn't doing the job in Angular. Is it because Angular is bulk updating the web component properties at the same time (settings loadingState and disabled at the same time?) or in a different order than expected (setting loadingState *before* disabled)? That would help me better understand your fix.

Likewise, I'm not sure I follow your unit tests updates and what they're attempting to do that's different than what they did before.

File: projects/core/src/button/button.element.ts [hunk 2]
```diff
@@ -83,34 +83,42 @@ export class CdsButton extends CdsBaseButton {
...
+    return super.disabled;
+  }
+
+  set disabled(value: boolean) {
+    // track disabled state set from outside, in order to restore value after button is back in default loading state
+    this._disabled_from_outside = value;
+
+    // do NOT propagate disabled state, unless we're in default loading state
+    // otherwise button can be enabled while being in a non-default loading state, e.g. loading
+    if (this.isDefaultLoadingState(this.loadingState)) {
```
Comment (ashleyryan) [inline projects/core/src/button/button.element.ts:96]: I was playing around with my own implementation loosely based off of yours and I was able to get around having this conditional logic here. I think we can always call `super` here and it still more or less works the same.

File: projects/core/src/button/button.element.spec.ts (hunk 1 above)
Comment (ashleyryan) [reply projects/core/src/button/button.element.spec.ts:306]: Can you keep the existing names and implementations for these unit tests. I think it's helpful to have tests where the implicit behavior happens - ie, no one ever set the disabled status before mucking with the loading states

File: projects/core/src/button/button.element.ts [hunk 3]
```diff
@@ -83,34 +83,42 @@ export class CdsButton extends CdsBaseButton {
...
-        this._disabled = this.disabled;
-      }
-
-      if (props.get('loadingState') !== undefined) {
-        if (this.isDefaultLoadingState(this.loadingState)) {
-          this.enableButton();
-        } else {
-          this.disableButton();
-        }
+      if (this.isDefaultLoadingState(this.loadingState)) {
```
Comment (ashleyryan) [inline projects/core/src/button/button.element.ts:111]: I'm trying to remember why, but I think there was a reason I had to have the undefined check here in the last PR. Otherwise I think disabled gets called if other prop values change - we only want to run this is loadingState is the prop changing

**Comment by ashleyryan** [comment]
This is causing some of the react snapshot/unit tests to fail and I'm investigating why.

**Comment by EndzeitBegins** [comment]
Thank you @ashleyryan for investing time looking into this issue.
I'm not quite certain whether there is something I can help you with at this moment.
Let me know if I can be of any help.

Should I change my pull-request in any way? Or do you simply take over the issue? 

---

I've found a slightly different solution that also passes the added tests. But still fails the aforementioned snapshot tests, as `lit` seems to add the attribute `disabled="true"` instead of former `disabled=""` to the DOM. However I don't know why that is the case or whether that is an actual problem or not.
However it no longer needs the conditional logic in the setter as you've lined out.

The idea is to basically use the `getter` to determine the `disabled` state both on the value stored in the parent and the `loadingState`:
```ts
  get disabled(): boolean {
    return super.disabled || !this.isDefaultLoadingState(this.loadingState);
  }

  set disabled(value: boolean) {
    super.disabled = value;
  }
```

Coupled with notifying `lit` about the potential update whenever the `loadingState` changes, this seems to have the same effect, without requiring an additional private property or explicit setting of the disabled state.

```ts
  update(props: PropertyValues<this>) {
    if (props.has('loadingState')) {
      // change in loadingState might also affect disabled state
      this.requestUpdate('disabled', super.disabled);

      if (this.isDefaultLoadingState(this.loadingState)) {
        this.style.removeProperty('width');
      } else {
        this.style.width = getElementWidth(this);
      }
    }

    super.update(props);
  }
```

With this approach, the override of `firstUpdated` can be dropped altogether.

**Comment by ashleyryan** [comment]
I can either take the PR over or you can try and make some of the changes you or I have suggested.

Why don't you try updating your commit to have the alternative implementation, I like that approach. We also need to update the commit message to include a link to the original issue. So update it to something like 

> fix(button): restore disabled state when loading state finishes
> 
> fixes #130

I've spent most of the day digging into why lit is doing that with boolean attributes and I've been talking to them. I've narrowed it down to something not working correctly when using a getter or setter vs a simple property. I think it's fine and I can update the react tests if necessary. But let's try the updated solution for now.

I'm also happy to take it over, if you'd like me to.

File: projects/core/src/button/button.element.ts (hunk 3 above)
Comment (ashleyryan) [reply projects/core/src/button/button.element.ts:111]: I remember why - props *always* had loadingState, but unless it's being changed, it's undefined. I found that without the undefined check, the else statement was being called on the initial render when `loadingState` was set, but the button hadn't been painted to the dom yet so the width was 0px. 

I added a test for this situation.

**Comment by ashleyryan** [comment]
I ended up taking it over @EndzeitBegins . I couldn't get your suggested approach to work - even though the tests were passing I was finding that in storybook when I was manipulating the component with 

```
document.querySelector('cds-button').loadingState = 'loading';
document.querySelector('cds-button').loadingState = 'default';
```

That it wasn't losing the disabled status for some reason.

**Comment by EndzeitBegins** [comment]
Thanks for looking into this and finding a solution that works @ashleyryan with the remaining use cases.

I hope my initial test cases were helpful in aiding the way. Looking forward to the merge!

File: projects/core/src/button/button.element.spec.ts [hunk 4]
```diff
@@ -310,6 +329,25 @@ describe('button element', () => {
...
+      // the width is only hardcoded when switching from default (with button text) to a loading status (with icon)
+      const testElement2 = await createTestElement(html`
+        <form>
+          <cds-button loading-state="loading">
+            <span>${placeholderText}</span>
+          </cds-button>
+        </form>
+      `);
+
+      component = testElement2.querySelector<CdsButton>('cds-button');
```
Comment (kevinbuhmann) [inline projects/core/src/button/button.element.spec.ts:343]: I think this should be `component2` to match `testElement2`.

File: projects/core/src/button/button.element.ts [hunk 5]
```diff
@@ -83,6 +83,21 @@ export class CdsButton extends CdsBaseButton {
   @property({ type: String })
   loadingState: keyof typeof ClrLoadingState = ClrLoadingState.default;
 
+  // when the loading state changes, the disabled state should be set to what the consumer manually set, if set
+  // the setter here should never be called in the component, call super instead
+  //  https://github.com/vmware-clarity/core/issues/129
+  private _disabled_externally = false;
```
Comment (kevinbuhmann) [inline projects/core/src/button/button.element.ts:89]: Snake case is usually not used in JS. So I think this should renamed to `_disabledExternally`.

File: projects/core/src/button/button.element.ts [hunk 6]
```diff
@@ -92,25 +107,26 @@ export class CdsButton extends CdsBaseButton {
...
-      }
-
-      if (props.get('loadingState') !== undefined) {
-        if (this.isDefaultLoadingState(this.loadingState)) {
-          this.enableButton();
-        } else {
-          this.disableButton();
-        }
+    if (props.has('loadingState') && props.get('loadingState') !== undefined) {
+      if (this.isDefaultLoadingState(this.loadingState)) {
```
Comment (kevinbuhmann) [inline projects/core/src/button/button.element.ts:111]: Why was this changed from `props.get('loadingState')` to accessing the property directly with `this.loadingState`?

File: projects/core/src/button/button.element.ts [hunk 7]
```diff
@@ -92,25 +107,26 @@ export class CdsButton extends CdsBaseButton {
...
-  private _disabled = false;
+  private disableButton() {
+    this.style.width = getElementWidth(this);
+    super.disabled = true;
+  }
+
+  private restoreButton() {
+    this.style.removeProperty('width');
+    super.disabled = this._disabled_externally;
+  }
```
Comment (kevinbuhmann) [inline projects/core/src/button/button.element.ts:129]: Can we put these private methods back where they were in the file? And rename `restoreButton` back to `enableButton`?

Comment (kevinbuhmann) [reply projects/core/src/button/button.element.ts:129]: Not only were these methods moved up in the file for some reason, they are now before a public method.

Comment (kevinbuhmann) [reply projects/core/src/button/button.element.ts]: > And rename `restoreButton` back to `enableButton`?

Nevermind, I see why you renamed it now.

**Review by kevinbuhmann — APPROVED** [review]
